feat: add parent directory to bufferline

mageowl opened the pull request:
Currently, it is impossible to distinguish between files that have the same name, but are in different directories. This PR introduces the option `editor.bufferline-show-parent` which prepends the first directory to the bufferline. By default, it is turned off.

Disabled:
![image](https://github.com/user-attachments/assets/9a9828e9-7623-43c6-941a-5f86011ccde6)

Enabled:
![image](https://github.com/user-attachments/assets/2fbe5225-e1fb-4b05-a5c5-e872ad1cb668)

mageowl committed:
add parent directory to bufferline

darshanCommits commented:
How would it handle the task in this situation

`modules/hardware/default.nix`
`hosts/hardware/default.nix`

Showing the full path make things very cluttered imo, esp for laptop users.
There has been a similar attempt made in the past #11836 which solves the cluttering issue but doesn't end up solving the root issue imo, which is at glance knowing what buffers are open. 

I think this needs discussion on what would be the best way to go around this. 
I think something like
1. Only *open buffers* with the same basename should trigger this behaviour. 

2. If buffers have the same basename, each buffer should look for the closest uniq parent directory. Like this.
> `modules/hardware/default.nix`
> `hosts/hardware/default.nix`
> `nix/default.nix`

3. Should have an option for
> disabled
> smart(above behaviour, naming could use some help)
> for all(for the large screen users)

poliorcetics commented:
Another solution is the "first character of path till different" algorithm, that avoids long names while still allowing for unicity:

```
modules/hardware/default.nix
hosts/hardware/default.nix
nix/default.nix
```

Becomes

```
m/h/default.nix
h/h/default.nix
n/default.nix
```

and if a `.name` is present, the shortening is done as `.n`, not just `.`

[starship](https://starship.rs/config/#directory) has a config like this for its `directory` module

mageowl committed:
add smart bufferline directories

mageowl committed:
remove single directory option

mageowl commented:
I agree with @darshanCommits. It makes more sense for this to only show up when needed. I think truncation works better when you are trying to get a sense for where you are in a project, like in a shell prompt, rather than when you are trying to distinguish between multiple files. (it would be the same for `helix-term/Cargo.toml` and `helix-lsp/Cargo.toml`) I went ahead and implemented a "smart" option. I don't think we need the always show, so I replaced it.

darshanCommits commented:
> I don't think we need the always show, so I replaced it.

I actually do have a use for it. When working with nix files, it's very common to have a `default.nix` and have the module information on the directory level.

Even if there's only single buffer open, it'd be desirable if, in such case, you could see more info like `ngnix/default.nix`

No idea if there are other languages with similar structuring (i think nextjs also have similar structuring)

mageowl committed:
add bufferline-directories=always back

mageowl commented:
Done!

mageowl committed:
Merge branch 'master' into master

mageowl committed:
Merge branch 'helix-editor:master' into master

mageowl committed:
Merge branch 'helix-editor:master' into master

mageowl committed:
fix(typst): highlight equal sign and lambda operator

mageowl committed:
Merge branch 'typst-equals-highlight'

mageowl committed:
fix incorrect query name

mageowl committed:
Merge branch 'typst-equals-highlight'

pheki commented:
@mageowl just to be sure, did you close this PR intentionally or was it a side-effect of e.g. renaming `master` to `main`?

mageowl commented:
This PR is a duplicate implementation of #13565
